### PR TITLE
JAMES-2097 Integration tests for RemoteDelivery partialSend

### PR DESCRIPTION
**Pull request by chibenwa**
I reproduce the bug reported in JAMES-2097.

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 1]
```diff
@@ -370,6 +382,57 @@ public void remoteDeliveryShouldNotDuplicateContentWhenSendPartial() throws Exce
...
+
+        given()
+            .get("/smtpMails")
+        .then()
+            .body("", hasSize(1))
+            .body("[0].from", is(FROM))
+            .body("[0].recipients", hasSize(1))
+            .body("[0].recipients[0]", is(RECIPIENT2))
+            .body("[0].message", containsString("subject: test"));
+        given(requestSpecification(mockSmtp2.getContainerIp()), RESPONSE_SPECIFICATION)
```
Comment (Arsnael) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: Add line break between calls

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 2]
```diff
@@ -370,6 +382,57 @@ public void remoteDeliveryShouldNotDuplicateContentWhenSendPartial() throws Exce
...
+    public void remoteDeliveryShouldNotDuplicateContentWhenSendPartialWhenFailover() throws Exception {
+        ImmutableList<InetAddress> addresses = ImmutableList.of(InetAddress.getByName(mockSmtp.getContainerIp()));
+        ImmutableList<String> mxs = ImmutableList.of(mockSmtp.getContainerIp(), mockSmtp2.getContainerIp());
+        ImmutableList<String> txtRecords = ImmutableList.of();
+
+        inMemoryDNSService.registerRecord(ANOTHER_DOMAIN, addresses, mxs, txtRecords)
+            .registerMxRecord(mockSmtp.getContainerIp(), mockSmtp.getContainerIp())
+            .registerMxRecord(mockSmtp2.getContainerIp(), mockSmtp2.getContainerIp());
+
+        with()
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: will be more clear if you declare a rest client for `mockstmp` and `mockstmp2` like:
```java
mockSmtp1Client = requestSpecification(mockSmtp1.getContainerIp());
mockSmtp2Client = requestSpecification(mockSmtp2.getContainerIp());

with(mockSmtp1Client)
            .body(ALWAYS_PARTIAL_RCPT_421_BEHAVIOR)
            .put("/smtpBehaviors");
```

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 3]
```diff
@@ -385,13 +448,13 @@ public void remoteDeliveryShouldNotDuplicateContentWhenSendPartial() throws Exce
                 .addProperty("sendpartial", "true"));
     }
 
-    private RequestSpecification requestSpecification() {
+    private RequestSpecification requestSpecification(String ip) {
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: I prefer to pass the `mockSmtp` instance here

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 4]
```diff
@@ -165,8 +177,8 @@ public void setUp() throws Exception {
             .addDomain(DEFAULT_DOMAIN)
             .addUser(FROM, PASSWORD);
 
-        RestAssured.requestSpecification = requestSpecification();
-        RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
+        RestAssured.requestSpecification = requestSpecification(mockSmtp.getContainerIp());
+        //RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: can you remove unused code?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 5]
```diff
@@ -140,13 +149,16 @@ private static MockSmtpBehaviors reject521Behavior(SMTPCommand smtpCommand) {
     public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
     @ClassRule
     public static DockerContainer mockSmtp = DockerContainer.fromName("chibenwa/mock-smtp-server")
-        .withLogConsumer(outputFrame -> LOGGER.debug(outputFrame.getUtf8String()));
+        .withLogConsumer(outputFrame -> LOGGER.debug("1: " + outputFrame.getUtf8String()));
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: a more clear prefix? `MockSmtpServer 1: ` ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 6]
```diff
@@ -62,11 +63,13 @@
 import org.junit.After;
 import org.junit.Before;
 import org.junit.ClassRule;
+import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: use guava?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 7]
```diff
@@ -370,6 +382,57 @@ public void remoteDeliveryShouldNotDuplicateContentWhenSendPartial() throws Exce
...
+            .body("[0].recipients[0]", is(RECIPIENT2))
+            .body("[0].message", containsString("subject: test"));
+        given(requestSpecification(mockSmtp2.getContainerIp()), RESPONSE_SPECIFICATION)
+            .get("/smtpMails")
+        .then()
+            .body("", hasSize(1))
+            .body("[0].from", is(FROM))
+            .body("[0].recipients", hasSize(1))
+            .body("[0].recipients[0]", is(RECIPIENT1))
+            .body("[0].message", containsString("subject: test"));
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: I can contribute a convenience MockSMTPClient with Feign

Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: Let's keep it simple?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java (hunk 2 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: `with` do not take arguments

`given` does

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java [hunk 8]
```diff
@@ -307,6 +328,48 @@ public void remoteDeliveryShouldRetryWhenDATA421() throws Exception {
...
+        messageSender.connect(LOCALHOST_IP, jamesServer.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .sendMessage(MailImpl.builder()
+                .name("name")
+                .sender(new MailAddress(FROM))
+                .addRecipient(RECIPIENT1)
+                .addRecipient(RECIPIENT2)
+                .mimeMessage(MimeMessageUtil.mimeMessageFromString(MIME_MESSAGE))
+                .build());
+
+        awaitAtMostOneMinute.untilAsserted(() -> given(requestSpecification(), RESPONSE_SPECIFICATION)
```
Comment (blackheaven) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: we tend to use `until` here

Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java]: I don't like `until` as when combined with rest-assured it makes things unessessarilly verbose.

Though I want my overall work to move forward hence I still contributed the fixup you ask for.

Do you think we should keep it?

**Comment by trantienduchn** [comment]
```
[a98c86dc5be4248c2cec512f0815896cb60d00ad] Error response from daemon: Could not find the file /spring/destination/james-server-app-3.4.0-SNAPSHOT-app.zip in container james-parent-test-james-2097-c41b2971-7732-4511-9d4f-b91fbddbcf61
```
passed mvn test

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**